easy_dispatch: support for writing dispatched functions

federkamm opened the pull request:
Code is tested on macOS 10.15.2 with (Apple) clang-11.0.0 on Intel i7-6700K (only).

AgnerF commented:
Thank you for your proposal.

Your code is very complicated to use. It is not clear to me what advantages your code has compared with the much simpler dispatch_example.cpp

I think that in simple cases, only one function needs to be dispatched. The dispatched function may call other static functions in the same module (compiled unit) compiled for the same instruction set. Dispatching every function will just make the code less efficient.

I am trying to avoid exception handling in VCL because it does not work well with vectors, see https://www.agner.org/optimize/nan_propagation.pdf even though the problems described herein do not apply to your case.

There is also a problem with copyright ownership. If VCL consists of different pieces with each their copyright owner, then it will be impossible to make any legal decisions without contacting all copyright owners. I have not found a good solution to this problem yet. It may be better to assign copyright ownership to an organization. I have asked gnu.org but they are not interested.

federkamm commented:
I removed the exceptions.  If no better implementation is found, the one with the lowest instruction set is returned.  If the CPU is not capable if it, it will likely raise a SIGILL.  But since then the rest of the program will likely not run on this CPU anyway, this might not be a real drawback.

I understand the problem with the copyright and the legal decisions.  I have to agree I felt a little bit funny when I added the Copyright and License note, but my main intend was to match the form of the other file headers and it felt wrong to place someones else name there without having asked before.  I would have released the code as Public Domain and you can use it like so.  Tell me, if you need anything more than this note, here.

I do understand that easy_dispatch.h is difficult to read but I don't understand why you say it would be difficult to use.  I thought the file easy_dispatch.cpp demonstrates that it is quite easy to use.  I understand that it might be quite some overkill if one wants to dispatch just a single function but I think that often enough it won't stay with one function and then the help of the three macros in easy_dispatch.h could be very appreciated by some programmers.

If one wants to call one dispatched function from another one can call VCL_DISPATCHED(fun)(...) instead of fun(...) to get the function name for the implementation (instruction set) and allow for inlining and other optimisations.  (Maybe a shorter macro name would be better here, e.g. VCL_DEF instead of VCL_DISPATCHED.)

AgnerF commented:
I have now added a new example dispatch_example2.cpp that uses different namespaces for the different versions. This eliminates the need for all the composite function names, and it also works with classes, objects, etc.

I don't think we need extra macros to handle this. I prefer to make it simple and transparent to the programmer.
